Can bump open firelocks with crowbar in active hand

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

https://user-images.githubusercontent.com/25089914/155260045-89aad25e-5b09-4889-9ad4-f28c49598814.mp4


If you bump into a firelock with a crowbar in your active hand (or as your active module as a cyborg), it will open as though you'd clicked on it.

~~Got most of this code from automatically mining bumped turfs.~~ I made it a component, although I'm not sure if it's any good or not.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Makes navigating firelock hell slightly quicker, without repeated clicking. They still stop you for a second if you have a crowbar and GBJ you if you don't.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: cacogen
qol: Firelocks will open when walked into with a crowbar or fireaxe in your active hand
qol: Resonator can be bump mined with (way less unwieldy than spam-clicking with the thing)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
Can bump open firelocks with crowbar in active hand

Razharas commented:
make this generic for any item on any bump target

cacogen commented:
I think that would lead to some undesirable interactions, like bumping into an airlock with a welder starting to fix it or weld it closed. 

I think it only makes sense in certain situations, like when you know the player wants to move onto the same turf as the atom after using a tool on it. And it's only really needed when it's likely they'll be doing it repeatedly (e.g. mining turfs, opening firelocks). Another case I can think of is opening depowered airlocks. 

I don't know how maintainers like oranges and LemonInTheDark are going to feel about this, though. I don't think having to click repeatedly is a desirable or necessary part of game balance but they may disagree. And it's made worse by lag. The click cooldown causes like one click to go through each second, and some of those are dropped due to lag. Movement is more responsive.

GuillaumePrata commented:
Bumping into an airlock will hold it open then?
Because if it is just the quick open that let it auto drops after a second, being smashed by firelocks again will get old real fast.

cacogen commented:
It just acts like a left-click. So it holds it open, yeah.

cacogen committed:
Stops chickens trying to open firelocks

SuperNovaa41 commented:
> I think that would lead to some undesirable interactions, like bumping into an airlock with a welder starting to fix it or weld it closed.
> 
> I think it only makes sense in certain situations, like when you know the player wants to move onto the same turf as the atom after using a tool on it. And it's only really needed when it's likely they'll be doing it repeatedly (e.g. mining turfs, opening firelocks). Another case I can think of is opening depowered airlocks.
> 
> I don't know how maintainers like oranges and LemonInTheDark are going to feel about this, though. I don't think having to click repeatedly is a desirable or necessary part of game balance but they may disagree. And it's made worse by lag. The click cooldown causes like one click to go through each second, and some of those are dropped due to lag. Movement is more responsive.

I believe he means turn it into an element, so it can be applied to any tool. so instead of copy pasting this code, you would just apply the element to the crowbar, with a callback to pry whenever the atom is a firelock (or something to that extent)

DrDiasyl commented:
Rebalance? This sounds like more of quality of life tho.

moocowswag commented:
Ya its definitely a QoL, its not like its letting you bypass click cooldowns or anything, infact I dont think clicking on a closed firelock even gives you a cooldown and youre able to just quickly spam click all of the ones near you open.

cacogen committed:
Tool_bump element

- Replaces code in Bumped for mineral turfs and firelocks with the element
- Adds the Bumped signal to firelocks (I don't know if this is allowed) because the fuckers were overriding Bumped() (how can you not bump something?)
- Fixes what I believe was a bug with mining where skill did not improve speed due to a cooldown that stopped mining to prevent message spam from bumping that did not factor skill

cacogen committed:
Makes click cooldown CD_MELEE to prevent people ATATATATA'ing firelocks

cacogen committed:
Minerals

- Makes it so tool bumping can require active handedness to work
- Makes it so tool bumping won't work with combat mode on to avoid autobludgeoning things
- Makes tool_bump cooldown only apply to bumps on the same atom by the same mob
- Makes bump opening firelocks require active handedness like it's supposed to
- Makes bump mining not require active handedness like it's supposed to
- Makes mining speeds in minerals.dm variables instead of magic numbers and uses the SECONDS define
- Uses cooldown defines for mining, which makes it so multiple people can try to mine the same tile simultaneously
- Resets cooldowns on failure to prevent mining being blocked when immediately trying again 
- Makes cooldowns on hand and tool mining the same length as the actions themselves to prevent doing it multiple times to the same turf simultaneously

cacogen commented:
I hope this gets reviewed now

cacogen committed:
Undoes changes to minerals.dm cooldowns

Changes for a different PR

timothymtorres commented:
This is a nice QoL change. Sometimes people are too slow to click on a firelock and move through it in time.  Whoever reviews this PR needs to remove the Balance label.

TheBonded commented:
this is going to save so much time and annoyances

cacogen commented:
I don't think it's ever going to be merged though...

TheBonded commented:
> I don't think it's ever going to be merged though...

ping the maintainers

LemonInTheDark commented on code/game/machinery/doors/firedoor.dm:
```diff
@@ -290,11 +294,12 @@
 		return
 	if(!density)
 		return ..()
+	SEND_SIGNAL(src, COMSIG_ATOM_BUMPED, movable)
```
this is really horrible you know

cacogen replied:
I don't understand the significance of it, to tell you the truth. It doesn't return parent by default, so how are you supposed to get all the way up the chain to where it calls BUMPED at the atom level? Do I have to make it call parent and then fix whatever unintended functionality arises from that?

LemonInTheDark replied:
![image](https://user-images.githubusercontent.com/58055496/156876936-eefa7afd-14bf-4d7d-9b75-88b0e81bc741.png)

cacogen replied:
I've gone through that and from what I can tell by calling parent the only issue that arises with firelocks is throwing your ID to open them which I've fixed in the code I'm about to test then push.

LemonInTheDark commented:
> Makes navigating firelock hell slightly quicker, without repeated clicking. They still stop you for a second if you have a crowbar and GBJ you if you don't.

do we really need to optimize away one click?

cacogen commented:
It's like you want me to add stat tracking for the number of firelocks crowbarred open by each player each round so we can see if on average it's a significant amount or not

cacogen committed:
Remove BUMPED signal from fire door's Bumped(), call parent instead

- Makes throwing an ID at a door to open it check if the door requires an ID before checking its access
- Override fire door's requiresID() to prevent using thrown IDs to magically open fire doors
- Fix both mine and Mothblocks' mistakes on fire door screen tips, expands them for different mob types
- Makes tool_behaviour tips only show to adjacent mobs
- Does some completely unrelated weird shit of providing feedback when making no damage paw attacks on machines (consistent with no damage messages)

cacogen committed:
Makes it work with mining resonator and fireaxe

cacogen committed:
Fikou told me ComponentInitialize() was deprecated

cacogen committed:
eurrrrrrgh

LemonInTheDark commented:
I don't need like, empirical data. I can imagine opening like 20 firelocks might be a bit annoying, but not everything that's annoying is bad. 
Fire alarms have had a ton of the edges sanded off recently, and I don't want to just make them not a problem, or not at least a bit frustrating. I don't like that pattern.

cacogen committed:
I like this element

cacogen committed:
hmmm ok

cacogen committed:
Undoes change to _machinery.dm

cacogen committed:
well shit

cacogen committed:
Undoes changes to firedoor screentips

optimumtact commented:
there does not appear to be any maintainer wanting to merge this.
